### PR TITLE
[Feature] Support Backup/Restore for external catalog

### LINKED ISSUE #52902: [Feature] Support Backup/Restore for external catalog
- [x] https://github.com/StarRocks/starrocks/pull/52895
In this PR, we support Backup/Restore for External Catalog. User can Backup All of or part of
External Catalog metadata into snapshot file and restore it into a new cluster.

The following summarizes the behavioral changes caused by this PR:
1. New syntax to support `Backup/Restore` for `External Catalog`, here is the new part for it:
```
BACKUP (ALL EXTERNAL CATALOGS | EXTERNAL CATALOG <catalog_name> [, ...]) SNAPSHOT <snapshot_name>
TO <repo_name> [PROPERTIES]

RESTORE SNAPSHOT <snapshot_name> FROM <repo_name>
[ ALL EXTERNAL CATALOGS | EXTERNAL CATALOG <catalog_name> AS <alias> [, ...] ]
[PROPERTIES]

CANCEL EXTERNAL CATALOG BACKUP

CANCEL EXTERNAL CATALOG RESTORE

```
2. For `EXTERNAL CATALOG` `Backup/Restore`, no db info and `ON` clause can be specified.
3. Only one `Backup/Restore` job can be executed at the same time in the cluster.

### PR DESCRIPTION
**Pull request by srlch**
In this PR, we support Backup/Restore for External Catalog. User can Backup All of or part of External Catalog metadata into snapshot file and restore it into a new cluster.

The following summarizes the behavioral changes caused by this PR:
1. New syntax to support `Backup/Restore` for `External Catalog`, here is the new part for it (old syntax is not shown here):
```
BACKUP { ALL EXTERNAL CATALOGS | EXTERNAL CATALOG[S] <catalog_name> [, ...] }
SNAPSHOT <snapshot_name> TO <repo_name>
[ PROPERTIES ]

RESTORE SNAPSHOT <snapshot_name> FROM <repo_name>
[ ALL EXTERNAL CATALOGS | EXTERNAL CATALOG[S] <catalog_name> AS <alias> [, ...] ]
[ PROPERTIES ]


-- CANCEL
CANCEL { BACKUP | RESTORE } FOR EXTERNAL CATALOG[S]
-- CANCEL { BACKUP | RESTORE } FROM [ DATABASE ] <db_name>
```
2. For `EXTERNAL CATALOG` `Backup/Restore`, no db info and `ON` clause can be specified.
3. Only one `Backup/Restore` job can be executed at the same time in the cluster.
4. `default_catalog` is not support.

Fixes #52902

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -1731,14 +1731,14 @@ privObjectTypePlural
 // ---------------------------------------- Backup Restore Statement ---------------------------------------------------
 
 backupStatement
-    : BACKUP (DATABASE dbName=identifier)?
+    : BACKUP (ALL EXTERNAL CATALOGS | EXTERNAL CATALOG identifierList)? (DATABASE dbName=identifier)?
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -3439,9 +3440,36 @@ private ParseNode parseBackupRestoreStatement(ParserRuleContext context) {
             backupContext = (StarRocksParser.BackupStatementContext) context;
         }
 
+        List<CatalogRef> externalCatalogRefs = new ArrayList<>();
+        boolean allExternalCatalog = backupContext != null ?
+                                     (backupContext.CATALOGS() != null) : (restoreContext.CATALOGS() != null);
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupHandler.java [hunk 3]
```diff
@@ -277,30 +284,49 @@ public void process(AbstractBackupStmt stmt) throws DdlException {
...
-        if (dbName == null) {
-            // if target dbName if null, use dbName in snapshot
-            dbName = jobInfo.dbName;
-        }
+            if (jobInfo.dbName == null || jobInfo.dbName.isEmpty()) {
+                // if jobInfo.dbName == null, means that this snapshot only contains external catalog info
+                if ((stmt.getOriginDbName() != null && !stmt.getOriginDbName().isEmpty()) ||
+                        stmt.getDbName() != null && !stmt.getDbName().isEmpty()) {
+                    ErrorReport.reportDdlException(ErrorCode.ERR_COMMON_ERROR,
+                                                   "Can not specify database for external catalog snapshot");
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/BackupHandler.java:293]: It seems that we should keep the code style here. All error msg for backup/restore is defined independently.

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupHandler.java [hunk 4]
```diff
@@ -124,6 +128,8 @@ public class BackupHandler extends FrontendDaemon implements Writable, MemoryTra
     // If the last job is finished, user can get the job info from repository. If the last job is cancelled,
     // user can get the error message before submitting the next one.
     // Use ConcurrentMap to get rid of locks.
+    // Backup/Restore job for external catalog using -1 to identify the job in dbIdToBackupOrRestoreJob
+    // which means that only one external catalog backup/restore job can be run in entire cluster
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/BackupHandler.java:132]: could define a constant FAKE_DB_ID  = -1 to make the code more clear

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 5]
```diff
@@ -1731,14 +1731,14 @@ privObjectTypePlural
...
     SNAPSHOT qualifiedName TO repoName=identifier
     (ON '(' backupRestoreObjectDesc (',' backupRestoreObjectDesc) * ')')?
     (PROPERTIES propertyList)?
     ;
 
 cancelBackupStatement
-    : CANCEL BACKUP ((FROM | IN) identifier)?
+    : CANCEL (EXTERNAL CATALOG)? BACKUP ((FROM | IN) identifier)?
```
[the remark this answers is not in the document]
Comment (xiangguangyxg) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: better to not add the new special sql.
could use cancel backup without from or in clause for external catalog ? same with restore.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 6]
```diff
@@ -1731,14 +1731,14 @@ privObjectTypePlural
 // ---------------------------------------- Backup Restore Statement ---------------------------------------------------
 
 backupStatement
-    : BACKUP (DATABASE dbName=identifier)?
+    : BACKUP (ALL EXTERNAL CATALOGS | EXTERNAL (CATALOG | CATALOGS) identifierList)? (DATABASE dbName=identifier)?
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1734]:
```suggestion
    : BACKUP (ALL EXTERNAL CATALOGS | EXTERNAL CATALOG identifier | EXTERNAL CATALOGS identifierList)? (DATABASE dbName=identifier)?
```

**Review by jaogoy — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by jaogoy — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**